src/ui/hooks/comments/useAddComment.tsx

```diff
@@ -29,67 +30,22 @@ export default function useAddComment() {
   return (comment: Comment) => {
     trackEvent("comments.create");
 
-    return addComment({
-      variables: {
-        input: {
-          ...omit(comment, ["id", "createdAt", "updatedAt", "replies", "user"]),
-          recordingId: comment.recordingId,
-        },
-      },
-      optimisticResponse: {
-        addComment: {
-          success: true,
-          comment: {
-            id: comment["id"],
-            __typename: "Comment",
+    return new Promise((resolve, reject) => {
+      addComment({
+        variables: {
+          input: {
+            ...omit(comment, ["id", "createdAt", "updatedAt", "isUnpublished", "replies", "user"]),
+            recordingId: comment.recordingId,
           },
-          __typename: "AddComment",
         },
-      },
-      update: (cache, { data }) => {
-        const commentId = data?.addComment?.comment?.id;
-        const cacheData = cache.readQuery<GetComments>({
-          query: GET_COMMENTS,
-          variables: { recordingId: comment.recordingId },
-        })!;
-        const {
-          viewer: {
-            user: { id: userId },
-          },
-        }: any = cache.readQuery({
-          query: GET_USER_ID,
-        });
-
-        const newComment = {
-          ...comment,
-          id: commentId,
-          primaryLabel: comment.primaryLabel || null,
-          secondaryLabel: comment.secondaryLabel || null,
-          user: {
-            id: userId,
-            name: comment.user.name,
-            picture: comment.user.picture,
-            __typename: "User",
-          },
-          __typename: "Comment",
-        };
-        const newData = {
-          ...cacheData,
-          recording: {
-            ...cacheData.recording,
-            comments: [
-              ...(cacheData.recording?.comments ?? []).filter(c => c.id !== commentId),
-              newComment,
-            ],
-          },
-        };
-
-        cache.writeQuery({
-          query: GET_COMMENTS,
-          variables: { recordingId: comment.recordingId },
-          data: newData,
-        });
-      },
+        onCompleted: (data: Comment | {}) => {
+          resolve(data);
+        },
+        onError: (error: ApolloError) => {
+          reject(error);
+        },
+        awaitRefetchQueries: true,
```

Comment:
We should not consider the operation complete (aka we shouldn't invoke our `onCompleted` handler) until the comment has been saved _and_ the list has been refetched. Otherwise the new comment will disappear and reappear a second or two later.

**TODO* This doesn't seem to work as expected. Most of the time it does, but some times there will still be a few seconds delay where the comment disappears and then pops back in. I'm not sure why yet.

Reply:
Maybe this is complicated by our long-polling for comments? Whenever I have the phenomenon of "it goes away and comes back a few seconds later" it's because my initial update is failing and then long-polling eventually fixes it.

Reply:
> it's because my initial update is failing and then long-polling eventually fixes it.

Can you describe this scenario a little more?

Why would an initial update fail silently with Apollo? Is that common/expected?